store/store.go

```diff
@@ -499,6 +506,67 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
+// notifyAndAdvance will notify waiters in heightSub and advance contiguousHead
+// based on already written headers.
+func (s *Store[H]) notifyAndAdvance(ctx context.Context, headers ...H) {
+	// always inform heightSub about new headers seen
+	for _, h := range headers {
+		s.heightSub.Notify(h.Height())
+	}
+
+	currHead := s.contiguousHead.Load()
+	if currHead != nil {
+		s.advanceContiguousHead(ctx, (*currHead).Height())
+	}
+}
+
+// advanceContiguousHead return a new highest contiguous height
+// or returns the given height if not found.
+func (s *Store[H]) advanceContiguousHead(ctx context.Context, currHeight uint64) uint64 {
+	// TODO(cristaloleg): benchmark this timeout or make it dynamic.
+	advCtx, advCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer advCancel()
+
+	prevHeight := currHeight
+	var newHead H
+	for {
+		h, err := s.getByHeight(advCtx, currHeight+1)
+		if err != nil {
+			break
+		}
+		newHead = h
+		currHeight++
+	}
+
+	if currHeight > prevHeight {
+		s.updateContiguousHead(ctx, newHead)
+	}
+	return currHeight
+}
+
+func (s *Store[H]) updateContiguousHead(ctx context.Context, newHead H) {
+	s.contiguousHead.Store(&newHead)
+	s.heightSub.SetHeight(newHead.Height())
+	log.Infow("new head", "height", newHead.Height(), "hash", newHead.Hash())
+	s.metrics.newHead(newHead.Height())
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	b, err := newHead.Hash().MarshalJSON()
+	if err != nil {
+		log.Errorw("cannot marshal new head",
+			"height", newHead.Height(), "hash", newHead.Hash(), "err", err)
+		return
+	}
+
+	if err := s.ds.Put(ctx, headKey, b); err != nil {
+		log.Errorw("cannot put new head",
+			"height", newHead.Height(), "hash", newHead.Hash(), "err", err)
+		return
+	}
```

Comment:
I realized we have a bug. This `Put` gets overwritten in `flushLoop` by `flush` on loc 449. There we write whatever last header there is in `batch` and, as we don't have adjacency, this is not correct anymore. 

We can trigger this by test that produces a gap in a `Store` and we restart it. The restarted Store is gonna have a wrong head.

The possible solutions is to remove headKey update there. Also, we could remove headKey update here and make the one in the flush to use `Head` method to get the correct head instead. The latter allows to keep writing things atomically in a single `Datastore.Batch` write.

Reply:
Interesting observation. Have dropped update here and updated `flush`.

The only 1 edge case is that for the 1st `flush` contiguous head might not be set.

Reply:
You can set contiguousHead before calling `flush` in `Init`

Reply:
Done. Luckily this didn't break anything.